From 6534fd83fb33dc4eb23692412fa5a3fe22155465 Mon Sep 17 00:00:00 2001
From: Shivani Raval <sraval@habana.ai>
Date: Fri, 11 Apr 2025 11:05:29 +0300
Subject: [PATCH] [JSW-31] reset: added jgs files

As of now added the same code as FS1 for
JGS asic build.

Change-Id: I21b99eb5563a0e450310a93a98103f100cad5f3d
---

diff --git a/arch/jgs/dev.c b/arch/jgs/dev.c
new file mode 100644
index 0000000..6b8df87
--- /dev/null
+++ b/arch/jgs/dev.c
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+ * (C) Copyright 2024 HabanaLabs, Ltd.
+ * All rights reserved.
+ *
+ */
+
+#include <zephyr/device.h>
+#include <hl_platform_defs.h>
+#include <jgs_reg_map.h>
+#include <psoc_boot_conf_regs.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#define LOG_LEVEL		CONFIG_LOG_DEFAULT_LEVEL
+#define FSM_GRAD_RST_DONE	0x13
+#define FSM_STATE_RETRIES	1000
+LOG_MODULE_REGISTER(intel_jgs_reset);
+struct k_work reset_work;
+
+static void fs1_rst_notify_cxl_arc(void)
+{
+	/* TODO */
+}
+
+static void fs1_rst_prepare_xha_reset(void)
+{
+	/* TODO */
+}
+
+static void fs1_rst_send_cxl_reset(void)
+{
+	/* TODO */
+}
+
+static void fs1_rst_isolation_cfg(bool en)
+{
+	/* TODO */
+}
+
+static void fs1_rst_issue_unit_rst(bool reset)
+{
+	/* TODO */
+}
+
+static void fs1_rst_work(struct k_work *work)
+{
+	uint32_t  retry = FSM_STATE_RETRIES;
+
+	/* Early reset notification to CXL */
+	fs1_rst_notify_cxl_arc();
+
+	/* Prepare xHA for reset */
+	fs1_rst_prepare_xha_reset();
+
+	/* Prepare CXL arc for reset */
+	fs1_rst_send_cxl_reset();
+
+	/* Set ISO for the blocks which undergo reset */
+	fs1_rst_isolation_cfg(true);
+
+	/* Issue unit reset to block under reset */
+	fs1_rst_issue_unit_rst(true);
+
+	/* wait for 0.5 usec */
+	k_busy_wait(500);
+
+	/* End unit reset */
+	fs1_rst_issue_unit_rst(false);
+
+	/* Clear ISO for the blocks which underwent reset */
+	fs1_rst_isolation_cfg(true);
+
+	/* set the BTM FSM state to FSM_GRAD_RST_DONE */
+	CFG_WREG32(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_BTM_FSM), FSM_GRAD_RST_DONE);
+
+	/* wait for BTM FSM state done 0x0 */
+	while (CFG_RREG32(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_BTM_FSM))) {
+		retry--;
+		k_busy_wait(1);
+		if (!retry) {
+			LOG_ERR("Timedout waiting for FSM state to be idle");
+			return;
+		}
+	}
+
+}
+
+__unused static void fs1_rst_callback(const struct device *dev, void *data)
+{
+	k_work_submit(&reset_work);
+}
+
+void reset_dev_init(void)
+{
+	/* TODO: Notify FSM about the resets being handled by reset ARC */
+
+	/* TODO: get the reset dev and register for call back */
+
+	k_work_init(&reset_work, fs1_rst_work);
+}
